### DISCUSSION TITLE
Training for 2D materials

### DISCUSSION THREAD

**Discussion by rick9402** [mir-group/allegro#38] [Q&A]
I am currently working on using Allegro for training ML force fields for 2D materials. However, after training four models with an increasing number of training samples (1k, 2k, 4k, 8k, total dataset contains 10k), I don't see any considerable reduction in the MAE for energies and forces, as I observed when I reproduced the calculations for LiPS system. Therefore, I would like to know whether Allegro/NequIP works for training on 2D materials or, maybe, I need to consider an additional flag when investigating 2D systems. 

In addition, I would like to ask whether NequIP/Allegro can use parallelisation on GPUs. If it is possible, please, could you tell me how to activate the training on multiple GPUs on the same GPU-node? I tried modifying the batch script in our cluster but it didn't work.

Finally, my third and last question is how to set up the training file for periodic systems, since so far I have been working with an npz file instead of xyz. This is because when setting up with an npz file I have the opportunity to put the key mapping "cell", while on xyz I have not been able to do so.

```
dataset_file_name: Periodic.npz                                                    # path to data set file
key_mapping:
  z: atomic_numbers                                                                # atomic species, integers
  E: total_energy                                                                  # total potential eneriges to train to
  F: forces                                                                        # atomic forces to train to
  R: pos                                                                           # raw atomic positions
  lattice: cell
npz_fixed_field_keys:                                                              # fields that are repeated across different examples
  - atomic_numbers

chemical_symbol_to_type:
  H: 0
  C: 1
  N: 2
```

**Comment by Linux-cpp-lisp** [comment]
Hi @rick9402 ,

Thanks for your interest in our methods!

There is no particular reason why the techniques would not work for 2D material. In principle it could affect the normalization, but I really doubt that would make a difference.  More relevant questions would be:   how similar is your new data to your existing data?  Does the training error go towards zero as you add more data?  Are you measuring validation error on a similar validation set or test error on an OOD test set?

Regarding multi GPU _training_, no this is not presently supported. See https://github.com/mir-group/nequip/issues/210.

> This is because when setting up with an npz file I have the opportunity to put the key mapping "cell", while on xyz I have not been able to do so.

extXYZ files have full (and better) support for periodic systems and are our recommendation for all use cases.  See `pbc` and `cell` of an `ase.Atoms` object and how they are written out by `ase.io.write` into XYZ files.

**Reply by rick9402** [reply]
Thanks for the reply. I have a question then, how should I set up the training file for training with extxyz? I mean, I have an extxyz file and I set it up like this for the periodic system, yet the training does yield good results, even compared with my npz training:

`ase option

dataset: ase
dataset_file_name:  Periodic.extxyz
ase_args:
   format: extxyz`

My extyz file has the following format:

15
Lattice="7.319041921 0.0 0.0 -3.659617155 6.338418779 0.0 0.0 0.0 20.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=3.5042041427755066 pbc="T T T"
C       -1.22172113       3.52105954       1.65768590      -2.96389797      -1.71446895      -0.00000000
C       -1.20985636       4.92402109       1.65768590      -2.94617444       1.70125260       0.00000000
C        2.50550765       2.77899586       1.65768590      -3.70228229       2.13745463       0.00000000
C        1.22072188       3.52074907       1.65768590       3.01879052      -1.74313873       0.00000000
C       -0.00076745       2.82935734       1.65768590       0.01836755      -3.38532551      -0.00000000
C        3.66087963       0.77848176       1.65768590      -0.00416315      -4.19591315      -0.00000000
C       -0.00080064       5.63579913       1.65768590       0.00252835       3.42422462      -0.00000000
C        1.20871531       4.92428529       1.65768590       2.94061719       1.67694353      -0.00000000
C        4.81569034       2.77870684       1.65768590       3.63178705       2.10167216      -0.00000000
N        2.46842047       1.42365212       1.65768590      -9.28823467      -5.36423773       0.00000000
N        4.85306303       1.42365872       1.65768590       9.30211967      -5.37037485      -0.00000000
N        3.66071609       3.48880608       1.65768590       0.00139325      10.72592840       0.00000000
H       -2.15202164       5.46795549       1.65768590      -3.77698674       2.18047709      -0.00000000
H        2.15019359       5.46959149       1.65768590       3.77175569       2.18405963       0.00000000
H       -0.00227418       1.74135852       1.65768590      -0.00561999      -4.35855375       0.00000000

**Reply by Linux-cpp-lisp** [reply]
Training results should be identical regardless of the input data format, unless your npz format setup was somehow setting different periodic boundary conditions.

The XYZ snippet you've posted looks fine at first glance...

I would suggest starting with the simplest possible test you can think of on your data (in-domain test set, simple small dataset, etc.) to verify the correctness of your data preparation workflow before trying to set up more complicated problems.

**Reply by rick9402** [reply]
Thanks for you answer. I will try to see what went wrong, since training it as extxyz gave wrong predicted energies vs, energies...

**Comment by Linux-cpp-lisp** [comment]
For future reference, multi-GPU (and other accelerations) for training are now available: https://nequip.readthedocs.io/en/latest/guide/advanced_training.html#distributed-data-parallel-training, https://arxiv.org/abs/2504.16068.
